# Module 10 Application

## Challenge: Crypto Clustering

In this Challenge, you’ll combine your financial Python programming skills with the new unsupervised learning skills that you acquired in this module.

The CSV file provided for this challenge contains price change data of cryptocurrencies in different periods.

The steps for this challenge are broken out into the following sections:

* Import the Data (provided in the starter code)
* Prepare the Data (provided in the starter code)
* Find the Best Value for `k` Using the Original Data
* Cluster Cryptocurrencies with K-means Using the Original Data
* Optimize Clusters with Principal Component Analysis
* Find the Best Value for `k` Using the PCA Data
* Cluster the Cryptocurrencies with K-means Using the PCA Data
* Visualize and Compare the Results

### Import the Data

This section imports the data into a new DataFrame. It follows these steps:

1. Read  the “crypto_market_data.csv” file from the Resources folder into a DataFrame, and use `index_col="coin_id"` to set the cryptocurrency name as the index. Review the DataFrame.

2. Generate the summary statistics, and use HvPlot to visualize your data to observe what your DataFrame contains.


> **Rewind:** The [Pandas`describe()`function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) generates summary statistics for a DataFrame. 

In [9]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [10]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    Path("Resources/crypto_market_data.csv"),
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [11]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [12]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

This section prepares the data before running the K-Means algorithm. It follows these steps:

1. Use the `StandardScaler` module from scikit-learn to normalize the CSV file data. This will require you to utilize the `fit_transform` function.

2. Create a DataFrame that contains the scaled data. Be sure to set the `coin_id` index from the original DataFrame as the index for the new DataFrame. Review the resulting DataFrame.


In [13]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data = StandardScaler().fit_transform(df_market_data)

In [14]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(
    scaled_data,
    columns=df_market_data.columns
)

# Copy the crypto names from the original data
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data

In this section, you will use the elbow method to find the best value for `k`.

1. Code the elbow method algorithm to find the best value for `k`. Use a range from 1 to 11. 

2. Plot a line chart with all the inertia values computed with the different values of `k` to visually identify the optimal value for `k`.

3. Answer the following question: What is the best value for `k`?

In [15]:
# Create a list with the number of k-values to try
# Use a range from 1 to 11
k = list(range(1,11))

In [16]:
# Create an empy list to store the inertia values
inertia = []

In [17]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    model = KMeans(n_clusters=i, random_state=0)
    model.fit(df_market_data)
    inertia.append(model.inertia_)

In [18]:
# Create a dictionary with the data to plot the Elbow curve
market_data_elbow = {
    "k": k,
    "inertia": inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_market_data_elbow = pd.DataFrame(market_data_elbow)

In [86]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
# create a new variable for the plot
df_market_data_elbow_plot = df_market_data_elbow.hvplot.line(
    x="k",
    y='inertia',
    title='Crytpocurrencies Market Data, Elbow Curve',
    xticks=k
).opts(yformatter='%.0f')

# display the plot
df_market_data_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: What is the best value for k?
**Question:** What is the best value for `k`?

**Answer:** Because the elbow begins at k=3, the best value for k will be 3.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In this section, you will use the K-Means algorithm with the best value for `k` found in the previous section to cluster the cryptocurrencies according to the price changes of cryptocurrencies provided.

1. Initialize the K-Means model with four clusters using the best value for `k`. 

2. Fit the K-Means model using the original data.

3. Predict the clusters to group the cryptocurrencies using the original data. View the resulting array of cluster values.

4. Create a copy of the original data and add a new column with the predicted clusters.

5. Create a scatter plot using hvPlot by setting `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. Color the graph points with the labels found using K-Means and add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.

In [28]:
# Initialize the K-Means model using the best value for k
model_k3 = KMeans(n_clusters=3)

In [29]:
# Fit the K-Means model using the scaled data
model_k3.fit(df_market_data)

KMeans(n_clusters=3)

In [31]:
# Predict the clusters to group the cryptocurrencies using the scaled data
market_data_clusters_k3 = model_k3.predict(df_market_data)

# View the resulting array of cluster values.
market_data_clusters_k3

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0],
      dtype=int32)

In [32]:
# Create a copy of the DataFrame
df_market_data_k3_predictions = df_market_data.copy()

In [33]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_k3_predictions["Clusters"] = market_data_clusters_k3

# Display sample data
df_market_data_k3_predictions

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                   1.08388                     7.60278   
ethereum                                  0.22392                    10.38134   
tether                                   -0.21173                     0.04935   
ripple                                   -0.37819                    -0.60926   
bitcoin-cash                              2.90585                    17.09717   
binancecoin                               2.10423                    12.85511   
chainlink                                -0.23935                    20.69459   
cardano                                   0.00322                    13.99302   
litecoin                                 -0.06341                     6.60221   
bitcoin-cash-sv                           0.92530                     3.29641   
crypto-com-chain                          0.61209                    -5.67151   
usd-coin                                 -0.17825                    -0.11871   
eos                                       0.14477                    -1.31177   
monero                                    0.42961                    15.78515   
tron                                      0.07647                     4.23886   
tezos                                    -0.67316                     8.95665   
okb                                      -2.72700                    -4.55389   
stellar                                  -1.00843                     2.07149   
cosmos                                   -0.95103                    16.08534   
cdai                                      0.21169                     0.05820   
neo                                       0.49302                     2.44243   
wrapped-bitcoin                           1.10231                     7.40537   
leo-token                                -0.13192                    -1.34886   
huobi-token                              -0.40818                     1.61798   
nem                                      -0.84990                    -0.85140   
binance-usd                              -0.10642                     0.04726   
iota                                      0.41996                     6.06830   
vechain                                   1.28766                    -1.76352   
zcash                                    -0.60897                    10.34780   
theta-token                              -4.56089                    -6.09456   
dash                                     -1.06006                     5.09387   
ethereum-classic                         -0.45950                     3.05209   
ethlend                                 -13.52786                     4.21266   
maker                                    -0.60285                     8.15400   
havven                                   -4.07216                     4.33651   
omisego                                   4.84033                     6.82985   
celsius-degree-token                      2.51323                     0.60354   
ontology                                 -1.35845                    -1.21399   
ftx-token                                 0.83416                     7.10438   
true-usd                                 -0.06197                     0.16642   
digibyte                                  2.97084                     0.67038   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                   6.57509   
ethereum                                  4.80849   
tether                                    0.00640   
ripple                                    2.24984   
bitcoin-cash                             14.75334   
binancecoin                               6.80688   
chainlink                                 9.30098   
cardano                                   5.55476   
litecoin

In [87]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# create a new variable for the plot
df_market_data_k3_predictions_plot = df_market_data_k3_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Clusters",
    hover_cols="coin_id",
    width=700,
    height=500,
    title="Cryptocurrencies Clustered by K=3"
)

# display the plot
df_market_data_k3_predictions_plot

:NdOverlay   [Clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis

In this section, you will perform a principal component analysis (PCA) and reduce the features to three principal components.

1. Create a PCA model instance and set `n_components=3`.

2. Use the PCA model to reduce to three principal components. View the first five rows of the DataFrame. 

3. Retrieve the explained variance to determine how much information can be attributed to each principal component.

4. Answer the following question: What is the total explained variance of the three principal components?

5. Create a new DataFrame with the PCA data. Be sure to set the `coin_id` index from the original DataFrame as the index for the new DataFrame. Review the resulting DataFrame.

In [38]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [43]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_market_data = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame. 
pca_market_data[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [44]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: What is the total explained variance of the three principal components?

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.3719856 + 0.34700813 + 0.17603793 = 0.89503166. The PCA contains about 90% of the original data.

In [46]:
# Create a new DataFrame with the PCA data.
# Note: The code for this step is provided for you

# Creating a DataFrame with the PCA data
df_pca_market_data = pd.DataFrame(
    pca_market_data,
    columns=["PC1","PC2","PC3"]
)

# Copy the crypto names from the original data
df_pca_market_data['coin_id'] = df_market_data.index

# Set the coinid column as index
df_pca_market_data = df_pca_market_data.set_index('coin_id')

# Display sample data
df_pca_market_data

PC1       PC2       PC3
coin_id                                           
bitcoin              -0.600667  0.842760  0.461595
ethereum             -0.458261  0.458466  0.952877
tether               -0.433070 -0.168126 -0.641752
ripple               -0.471835 -0.222660 -0.479053
bitcoin-cash         -1.157800  2.041209  1.859715
binancecoin          -0.516534  1.388377  0.804071
chainlink            -0.450711  0.517699  2.846143
cardano              -0.345600  0.729439  1.478013
litecoin             -0.649468  0.432165  0.600303
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653
crypto-com-chain     -0.248198 -1.376252 -1.462026
usd-coin             -0.438408 -0.175337 -0.663388
eos                  -0.693425 -0.473815 -0.527597
monero                0.060499  2.909404  1.498571
tron                 -0.393352 -0.108192 -0.012756
tezos                -0.796176 -0.494409  1.082812
okb                   0.064075 -1.269825 -1.098829
stellar              -0.489015 -0.732719 -0.062543
cosmos               -0.306272  0.703415  1.714224
cdai                 -0.513528 -0.142802 -0.656566
neo                  -0.362120 -0.986914 -0.728752
wrapped-bitcoin      -0.604265  0.827398  0.439316
leo-token            -0.413296 -0.674115 -1.076628
huobi-token          -0.407483 -0.212507 -0.351426
nem                   0.608974  0.563532 -1.148742
binance-usd          -0.450211 -0.151019 -0.647401
iota                 -0.764665 -0.517886  0.204990
vechain              -0.556315 -1.938209 -1.261776
zcash                -0.425147  0.492976  1.058048
theta-token           2.676868 -0.013954 -1.965207
dash                 -0.613923 -0.479337  0.339565
ethereum-classic     -0.579924 -0.356334 -0.114942
ethlend               8.089018 -3.896891  2.301382
maker                -0.389045  0.165041  0.379414
havven                0.865762 -2.261882  0.275583
omisego               0.111675  0.428316 -1.205398
celsius-degree-token  4.792395  6.767679 -1.986985
ontology             -0.632355 -2.108117 -0.652227
ftx-token            -0.593142  0.021485  0.209911
true-usd             -0.458131 -0.135734 -0.635284
digibyte             -0.297910 -0.191126 -0.909602

---

### Find the Best Value for k Using the PCA Data

In this section, you will use the elbow method to find the best value for `k` using the PCA data.

1. Code the elbow method algorithm and use the PCA data to find the best value for `k`. Use a range from 1 to 11. 

2. Plot a line chart with all the inertia values computed with the different values of `k` to visually identify the optimal value for `k`.

3. Answer the following questions: What is the best value for k when using the PCA data? Does it differ from the best k value found using the original data?

In [52]:
# Create a list with the number of k-values to try
# Use a range from 1 to 11
k = list(range(1,11))

In [53]:
# Create an empy list to store the inertia values
inertia = []

In [54]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    model = KMeans(n_clusters=i, random_state=0)
    model.fit(df_market_data)
    inertia.append(model.inertia_)

In [55]:
# Create a dictionary with the data to plot the Elbow curve
pca_market_data_elbow = {
    "k": k,
    "inertia": inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_pca_market_data_elbow = pd.DataFrame(pca_market_data_elbow)

In [88]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
# create a new variable for the plot
df_pca_market_data_elbow_plot = df_pca_market_data_elbow.hvplot.line(
    x="k",
    y='inertia',
    title='Crytpocurrencies PCA Market Data, Elbow Curve',
    xticks=k
).opts(yformatter='%.0f')

# display the plot
df_pca_market_data_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following questions: What is the best value for k when using the PCA data? Does it differ from the best k value found using the original data?
* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The elbow curse is starting at k=3 so the best value for k will be 3.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** This is the same value we found for the original data, k=3.

---

### Cluster Cryptocurrencies with K-means Using the PCA Data

In this section, you will use the PCA data and the K-Means algorithm with the best value for `k` found in the previous section to cluster the cryptocurrencies according to the principal components.

1. Initialize the K-Means model with four clusters using the best value for `k`. 

2. Fit the K-Means model using the PCA data.

3. Predict the clusters to group the cryptocurrencies using the PCA data. View the resulting array of cluster values.

4. Add a new column to the DataFrame with the PCA data to store the predicted clusters.

5. Create a scatter plot using hvPlot by setting `x="PC1"` and `y="PC2"`. Color the graph points with the labels found using K-Means and add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.

In [60]:
# Initialize the K-Means model using the best value for k
model_pca_k3 = KMeans(n_clusters=3)

In [61]:
# Fit the K-Means model using the PCA data
model_pca_k3.fit(df_pca_market_data)

KMeans(n_clusters=3)

In [62]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_market_data_clusters_k3 = model_pca_k3.predict(df_pca_market_data)

# View the resulting array of cluster values.
pca_market_data_clusters_k3

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0],
      dtype=int32)

In [63]:
# Create a copy of the DataFrame with the PCA data
df_pca_market_data_k3_predictions = df_pca_market_data.copy()

# Add a new column to the DataFrame with the predicted clusters
df_pca_market_data_k3_predictions["Clusters"] = pca_market_data_clusters_k3

# Display sample data
df_pca_market_data_k3_predictions

PC1       PC2       PC3  Clusters
coin_id                                                     
bitcoin              -0.600667  0.842760  0.461595         0
ethereum             -0.458261  0.458466  0.952877         0
tether               -0.433070 -0.168126 -0.641752         0
ripple               -0.471835 -0.222660 -0.479053         0
bitcoin-cash         -1.157800  2.041209  1.859715         0
binancecoin          -0.516534  1.388377  0.804071         0
chainlink            -0.450711  0.517699  2.846143         0
cardano              -0.345600  0.729439  1.478013         0
litecoin             -0.649468  0.432165  0.600303         0
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653         0
crypto-com-chain     -0.248198 -1.376252 -1.462026         0
usd-coin             -0.438408 -0.175337 -0.663388         0
eos                  -0.693425 -0.473815 -0.527597         0
monero                0.060499  2.909404  1.498571         0
tron                 -0.393352 -0.108192 -0.012756         0
tezos                -0.796176 -0.494409  1.082812         0
okb                   0.064075 -1.269825 -1.098829         0
stellar              -0.489015 -0.732719 -0.062543         0
cosmos               -0.306272  0.703415  1.714224         0
cdai                 -0.513528 -0.142802 -0.656566         0
neo                  -0.362120 -0.986914 -0.728752         0
wrapped-bitcoin      -0.604265  0.827398  0.439316         0
leo-token            -0.413296 -0.674115 -1.076628         0
huobi-token          -0.407483 -0.212507 -0.351426         0
nem                   0.608974  0.563532 -1.148742         0
binance-usd          -0.450211 -0.151019 -0.647401         0
iota                 -0.764665 -0.517886  0.204990         0
vechain              -0.556315 -1.938209 -1.261776         0
zcash                -0.425147  0.492976  1.058048         0
theta-token           2.676868 -0.013954 -1.965207         0
dash                 -0.613923 -0.479337  0.339565         0
ethereum-classic     -0.579924 -0.356334 -0.114942         0
ethlend               8.089018 -3.896891  2.301382         1
maker                -0.389045  0.165041  0.379414         0
havven                0.865762 -2.261882  0.275583         0
omisego               0.111675  0.428316 -1.205398         0
celsius-degree-token  4.792395  6.767679 -1.986985         2
ontology             -0.632355 -2.108117 -0.652227         0
ftx-token            -0.593142  0.021485  0.209911         0
true-usd             -0.458131 -0.135734 -0.635284         0
digibyte             -0.297910 -0.191126 -0.909602         0

In [89]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# create a variable for the plot
df_pca_market_data_k3_predictions_plot = df_pca_market_data_k3_predictions.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="Clusters",
    hover_cols="coin_id",
    width=700,
    height=500,
    title="Cryptocurrencies PCA Clustered by K=3"
)

# display the plot
df_pca_market_data_k3_predictions_plot

:NdOverlay   [Clusters]
   :Scatter   [PC1]   (PC2,coin_id)

---

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

1. Create a composite plot using hvPlot and the plus (`+`) operator to contrast the Elbow Curve that you created to find the best value for `k` with the original and the PCA data.

2. Create a composite plot using hvPlot and the plus (`+`) operator to contrast the cryptocurrencies clusters using the original and the PCA data.

3. Answer the following question: After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

> **Rewind:** Back in Lesson 3 of Module 6, you learned how to create composite plots. You can look at that lesson to review how to make these plots; also, you can check [the hvPlot documentation](https://holoviz.org/tutorial/Composing_Plots.html).

In [91]:
# create a new Composite plot to contrast the Elbow curves
elbow_composite_plot = df_market_data_elbow_plot + df_pca_market_data_elbow_plot

# display the plot
elbow_composite_plot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [92]:
# create a new Composite plot to contrast the clusters
cluster_composite_plot = df_market_data_k3_predictions_plot + df_pca_market_data_k3_predictions_plot

# display the plot
cluster_composite_plot

:Layout
   .NdOverlay.I  :NdOverlay   [Clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [Clusters]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The elbow curve plots are very similar, but in both curves k=3 is the obvious best value. The cluster plots are very different! In the PCA plot, the clusters are more tightly grouped together--more obvious clusters. It's clear that using fewer features to cluster the data provides a much more detailed picture of the data.